### PR TITLE
Refactor the IntToBoolConverter to use the BaseConverter

### PR DESCRIPTION
**Pull request by bijington**
### Description of Change ###

 

Refactors the `IntToBoolConverter` to use the `BaseConverter`.

 ### Linked Issues ###
 

n/a

 ### PR Checklist ###
 
 - [ ] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [ ] Has tests (if omitted, state reason in description)
 - [ ] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pulls/XYZ 

 ### Additional information ###

### CODE REVIEW

File: src/CommunityToolkit.Maui/Converters/IntToBoolConverter.shared.cs [hunk 1]
```diff
@@ -5,40 +5,21 @@
...
-		if (value is bool result)
-		{
-			return result ? 1 : 0;
-		}
-
-		throw new ArgumentException("Value is not a valid boolean", nameof(value));
-	}
+	/// <param name="value">The value to convert.</param>
+	/// <returns>`1` if the supplied `value` is `true` and `0` otherwise.</returns>
+	public abstract int ConvertBackTo(bool value) => value ? 1 : 0;
```
Comment (brminnick) [inline src/CommunityToolkit.Maui/Converters/IntToBoolConverter.shared.cs:24]: I assume `abstract` is a typo since it's also implemented?

**Review by brminnick — CHANGES_REQUESTED** [review]

**Comment by brminnick** [comment]
Adding the `blocked` tag until https://github.com/CommunityToolkit/Maui/pull/325 is merged

File: src/CommunityToolkit.Maui.UnitTests/Converters/IntToBoolConverter_Tests.cs [hunk 2]
```diff
@@ -13,9 +13,11 @@ public void IndexToArrayConverter(int value, bool expectedResult)
 	{
 		var intToBoolConverter = new IntToBoolConverter();
 
-		var result = (bool)intToBoolConverter.Convert(value, typeof(bool), null, CultureInfo.CurrentCulture);
+		var result = intToBoolConverter.Convert(value, typeof(bool), null, CultureInfo.CurrentCulture);
+		var typedResult = intToBoolConverter.ConvertFrom(value);
 
 		Assert.Equal(result, expectedResult);
+		Assert.Equal(typedResult, expectedResult);
```
Comment (brminnick) [inline src/CommunityToolkit.Maui.UnitTests/Converters/IntToBoolConverter_Tests.cs:20]: Let's swap these `Assert.Equal` parameters around.

It's kinda trivial, but when the `Assert` fails, it'll print "...Expected: [first parameter] Actual: [second parameter]" which can be confusing if the parameters are reversed

```suggestion
		Assert.Equal(expectedResult, result);
		Assert.Equal(expectedResult, typedResult);
```

File: src/CommunityToolkit.Maui.UnitTests/Converters/IntToBoolConverter_Tests.cs [hunk 3]
```diff
@@ -25,16 +27,17 @@ public void IndexToArrayConverterBack(bool value, int expectedResult)
 	{
 		var intToBoolConverter = new IntToBoolConverter();
 
-		var result = (int)intToBoolConverter.ConvertBack(value, typeof(bool), null, CultureInfo.CurrentCulture);
+		var result = intToBoolConverter.ConvertBack(value, typeof(bool), null, CultureInfo.CurrentCulture);
+		var typedResult = intToBoolConverter.ConvertBackTo(value);
 
 		Assert.Equal(result, expectedResult);
+		Assert.Equal(typedResult, expectedResult);
```
Comment (brminnick) [inline src/CommunityToolkit.Maui.UnitTests/Converters/IntToBoolConverter_Tests.cs:34]: Let's swap these `Assert.Equal` parameters around.

It's kinda trivial, but when the `Assert` fails, it'll print "...Expected: [first parameter] Actual: [second parameter]" which can be confusing if the parameters are reversed.

```suggestion
		Assert.Equal(expectedResult, result);
		Assert.Equal(expectedResult, typedResult);
```

File: src/CommunityToolkit.Maui.UnitTests/Converters/IntToBoolConverter_Tests.cs [hunk 4]
```diff
@@ -25,16 +27,17 @@ public void IndexToArrayConverterBack(bool value, int expectedResult)
...
 	}
 
 	[Theory]
 	[InlineData(2.5)]
 	[InlineData("")]
-	[InlineData(null)]
-	public void InValidConverterValuesThrowArgumenException(object value)
+	public void InValidConverterValuesThrowArgumentException(object value)
```
Comment (brminnick) [inline src/CommunityToolkit.Maui.UnitTests/Converters/IntToBoolConverter_Tests.cs:40]: Let's fix this small typo from a previous PR

```suggestion
	public void InvalidConverterValuesThrowArgumentException(object value)
```

File: src/CommunityToolkit.Maui.UnitTests/Converters/IntToBoolConverter_Tests.cs [hunk 5]
```diff
@@ -44,9 +47,16 @@ public void InValidConverterValuesThrowArgumenException(object value)
 	[InlineData(2.5)]
 	[InlineData("")]
 	[InlineData(null)]
-	public void InValidConverterBackValuesThrowArgumenException(object value)
+	public void InValidConverterBackValuesThrowArgumentException(object value)
```
Comment (brminnick) [inline src/CommunityToolkit.Maui.UnitTests/Converters/IntToBoolConverter_Tests.cs:50]: Let's fix this small typo from a previous PR
```suggestion
	public void InvalidConverterBackValuesThrowArgumentException(object value)
```

**Review by brminnick — CHANGES_REQUESTED** [review]
Thanks Shaun!

Could you also fix the Test Method names?
- `IndexToArrayConverter` -> `IntToBoolConverter`
- `IndexToArrayConverterBack` -> `IntToBoolConverterBack`

**Review by brminnick — APPROVED** [review]

**Comment by bijington** [comment]
@brminnick rhanks for the fix. I haven't removed the extensions around asserting the Excoetional conditions. What are your thoughts?

**Review by brminnick — DISMISSED** [review]

**Comment by brminnick** [comment]
Hey Shaun! I went ahead and pulled out the converter assert extension methods. I think it's best to keep the Unit Tests as simple/dumb as possible to make it easier for us to understand why unit tests are failing when writing future PRs.

For example, when Vincent Hoogendoorn wrote the unit tests for the Markup library, he made them very complex and now it's tough to understand what a Unit Test is doing and, worse, to understand why a Unit Test is failing: https://github.com/CommunityToolkit/Maui.Markup/blob/main/src/CommunityToolkit.Maui.Markup.UnitTests/BindableObjectExtensionsTests.cs

**Comment by bijington** [comment]
No problem, I get that argument and missed you had removed them. I wanted to make sure before this went in 😃

**Review by brminnick — APPROVED** [review]

**Pull request merged**